docs/deployment-design.md

```diff
@@ -0,0 +1,57 @@
+# Design doc: Model deployment pipeline
+
+This doc represents a proposal for a simple CI/CD pipeline that we can use to deploy residential models and run experiments on them.
+
+## Background
+
+At a high level, our **existing process** for experimenting with changes to our models looks like this:
+
+* Models run on an on-prem server
+* Data scientists trigger model runs by SSHing into the server and running shell commands from cloned copies of this repo
+* Model artifacts are saved using DVC
+* Data scientists commit corresponding changes to model code after their experiment runs prove successful  
+
+This process has the advantage of being simple, easy to maintain, and cheap; as a result it has been useful to our team during the recent past when we only had a few data scientists on staff and they needed to focus most of their effort on building a new model from the ground up. However, some of its **limitations** are becoming apparent as our team scales up and begins to expand our focus:
+
+* Our on-prem server only has enough resources to run one model at a time, so only one data scientist may be running modeling experiments at a given time
+  * Further, our server has no notion of a job queue, so a data scientist who is waiting to run a model must notice that a previous run has completed and initiate their run manually 
+* Our on-prem server does not have a GPU, so it can't make use of GPU-accelerated libraries like XGBoost
+* Model runs are decoupled from version control changes, so data scientists have to remember to commit their changes correctly
+* Results of model runs are not easily accessible to PR reviewers
+
+The design described below aims to remove these limitations while retaining as much simplicity, maintainability, and affordability as possible.
+
+## Requirements
+
+At a high level, a model deployment pipeline should:
+
+* Integrate with our existing cloud infrastructure (GitHub and AWS)
+* Trigger model runs from pull request branches
+* Require code authors to approve model runs before they are initiated
+* Run the model on ephemeral, cheap, and isolated cloud infrastructure
+* Run multiple model runs simultaneously on separate hardware
+* Report model statistics back to the pull request that triggered a run
+
+## Design
+
+Here is a rough sketch of a new model deployment pipeline:
+
+* Define a new workflow, `run-model.yaml`, that runs on:
+  * Every commit to every pull request against the main branch
```

Comment:
One possible gotcha here: in the past, we've done each year's model development on a long lived branch associated with that year, e.g. 2024 will be `2024-assessment-year`. The goal here was to keep the master branch a relatively stable place for people to see "finished" models/documentation.

I'm open to changing this pattern if it would complicate things a lot. Maybe we just put a big WIP banner on the README until 2024 is finished.

Reply:
That should be fine! I think we can just set the workflow to run on every pull request, rather than restricting it to PRs opened against the main branch.